File: Reverse_Linked_List.cpp
/**
 * Definition of ListNode
 * 
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 * 
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: The new head of reversed linked list.
     */
    ListNode *reverse(ListNode *head) {
        // write your code here
        if (!head) return head;
        ListNode *left = head, *right = head->next;
        if (!right) return head;
        left->next = nullptr;
        while (right->next){
            auto temp = right;
            right = right->next;
            temp->next = left;
            left = temp;
        }
        right->next = left;
        return right;
    }
};

File: Reverse Linked List.py
"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: You should return the head of the reversed linked list. 
                  Reverse it in-place.
    """
    def reverse(self, head):
        # write your code here
        # dummy = ListNode(-1)
        # cur = dummy
        # while head:
        #     tmp = head.next
        #     head.next = cur.next
        #     cur.next = head
        #     head = tmp
        # return dummy.next
        
        cur = None
        while head:
            tmp = head.next
            head.next = cur
            cur = head
            head = tmp
        return cur